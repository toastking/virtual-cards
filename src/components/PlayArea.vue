<template>
  <section>
    <div class="playing-card-container container box">
      <div class="level">
        <div
          class="level-item remaining-deck"
          v-if="!gameOver"
          v-on:click="turnHandler()"
        >
          <vue-playing-card cover></vue-playing-card>
        </div>
        <div class="level-item" v-for="deck of decks" :key="deck.id">
          <transition
            tag="div"
            name="playingcard"
            enter-active-class="animate__animated animate__flipInY"
          >
            <vue-playing-card
              v-if="!!deck.currentCard"
              :signature="deck.currentCard"
              :key="deck.currentCard"
            ></vue-playing-card>
          </transition>
        </div>
      </div>
    </div>
  </section>
</template>

<script lang="ts">
import { State } from '@/store';
import { throttle } from 'lodash';
import Vue from 'vue';
import { mapActions, mapGetters, mapState } from 'vuex';
export default Vue.extend({
  created() {
    this.turnHandler = throttle(() => {
      if ((this.isYourTurn as boolean) && !(this.turnIsLoading as boolean)) {
        this.doTurn();
      }
    }, 1000);
  },
  computed: {
    ...mapGetters(['gameOver', 'decks', 'isYourTurn', 'turnIsLoading']),
  },
  methods: { ...mapActions(['doTurn']), turnHandler() {} },
});
</script>

<style scoped>
.remaining-deck {
  cursor: pointer;
}
</style>
