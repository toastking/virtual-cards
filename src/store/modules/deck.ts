import { Module } from 'vuex';
import { State } from '@/store';
import { firestoreAction } from 'vuexfire';
import { db } from '@/db';

export interface Deck {
  drawnCards: string[];
  currentCard?: string;
  id?: string; // Id from firestore
}

export interface DeckState {
  decks: Deck[];
  currentDeckIndex: number;
}

const baseDeck = createInitialDeck();

export const DeckModule: Module<DeckState, State> = {
  state: () => ({ decks: [], currentDeckIndex: 0 }),
  actions: {
    /** Draw a random card from a deck in the game */
    drawCard({ rootState, getters }) {
      // Get the drawn cards for the current deck
      const availableCards: string[] = getters.cardsStillLeft;
      const cardIndex = Math.floor(Math.random() * availableCards.length);
      const drawnCard = availableCards[cardIndex];
      const currentDeck: Deck = getters.currentDeck;
      const updatedDrawnCards = currentDeck.drawnCards.concat(drawnCard);

      // update the Deck value in firestore
      const { gameId } = rootState.game;
      if (gameId) {
        const fieldsToUpdate: Partial<Deck> = {
          currentCard: drawnCard,
          drawnCards: updatedDrawnCards,
        };
        return db
          .collection('games')
          .doc(gameId)
          .collection('decks')
          .doc(currentDeck.id)
          .update(fieldsToUpdate);
      }
    },
    /** Add a deck to firebase */
    addDeck({ rootState }) {
      const { gameId } = rootState.game;
      if (gameId) {
        const newDeck: Deck = { drawnCards: [] };
        return db
          .collection('games')
          .doc(gameId)
          .collection('decks')
          .add(newDeck);
      }
    },
    setupDeckBinding: firestoreAction(({ rootState, bindFirestoreRef }) => {
      const { gameId } = rootState.game;
      if (gameId) {
        return bindFirestoreRef(
          'decks',
          db
            .collection('games')
            .doc(gameId)
            .collection('decks')
        );
      }
    }),
  },
  getters: {
    /** Calculate the cards still left to draw in a given deck */
    cardsStillLeft(state): string[] {
      const drawnCards = new Set(
        state.decks[state.currentDeckIndex].drawnCards
      );
      return baseDeck.filter(card => !drawnCards.has(card));
    },
    /** Get the current deck we're working on */
    currentDeck(state): Deck {
      return state.decks[state.currentDeckIndex];
    },
  },
};

/** Make the initial deck value */
function createInitialDeck(): string[] {
  const ranks = [
    'A',
    '2',
    '3',
    '4',
    '5',
    '6',
    '7',
    '8',
    '9',
    'T',
    'J',
    'Q',
    'K',
  ];
  const suits = ['c', 'd', 'h', 's'];
  const deck: string[] = [];

  for (const suit of suits) {
    for (const rank of ranks) {
      deck.push(`${rank}${suit}`);
    }
  }
  return deck;
}
