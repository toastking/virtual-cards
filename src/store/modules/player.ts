import { Module } from 'vuex';
import { db } from '@/db';
import { State } from '@/store';
import { firestoreAction } from 'vuexfire';

export interface Player {
  name: string;
  id?: string;
}

export interface PlayerState {
  players: [];
}

/** Module to handle player state, like who's in the game and who is the current player. */
export const PlayerModule: Module<PlayerState, State> = {
  state: () => ({ currentPlayerId: null, players: [] }),
  actions: {
    /** Add a player to the firebase store */
    async addPlayer({ rootState }, player: Player) {
      const { gameId } = rootState.game;
      if (gameId) {
        return db
          .collection('games')
          .doc(gameId)
          .collection('players')
          .add(player);
      }
    },
    /** Sets up the firebase binding for the players list  */
    setupPlayerBinding: firestoreAction(
      async ({ bindFirestoreRef, rootState }) => {
        const { gameId } = rootState.game;
        if (gameId) {
          return bindFirestoreRef(
            'players',
            db
              .collection('games')
              .doc(gameId)
              .collection('players')
          );
        }
      }
    ),
  },
};
