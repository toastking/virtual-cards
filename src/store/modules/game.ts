import { db } from '@/db';
import router from '@/router';
import { State } from '@/store';
import { Module } from 'vuex';
import { firestoreAction } from 'vuexfire';
import { Player } from './player';

export enum LoadingStatus {
  NOT_STARTED,
  LOADING,
  OK,
  ERROR,
}

/** Represents the general state of a card game */
export interface Game {
  currentPlayer: string | null;
  gameCompleted: boolean;
  gameStarted: boolean;
}

export interface GameState {
  gameId: string | null;
  gameLoadingStatus: LoadingStatus;
  game: Game;
  // Loading for the request for the turn
  turnLoadingState: LoadingStatus;
}

/** Module to handle game state. This is things like the game ID and the completion state. */
export const GameModule: Module<GameState, State> = {
  state: () => ({
    gameId: null,
    gameLoadingStatus: LoadingStatus.NOT_STARTED,
    turnLoadingState: LoadingStatus.OK,
    game: {
      currentPlayer: null,
      gameCompleted: false,
      gameStarted: false,
    },
  }),
  mutations: {
    createGameSuccess(state, payload: { newGameId: string }) {
      state.gameId = payload.newGameId;
      state.gameLoadingStatus = LoadingStatus.OK;
    },
    turnRequestStarted(state) {
      state.turnLoadingState = LoadingStatus.LOADING;
    },
    turnRequestDone(state) {
      state.turnLoadingState = LoadingStatus.OK;
    },
  },
  actions: {
    /** Create a new game object and route to the lobby */
    async createGame({ dispatch }, payload: { hostPlayerName: string }) {
      const dummyGame: Game = {
        currentPlayer: null,
        gameCompleted: false,
        gameStarted: false,
      };
      const newGame = await db.collection('games').add(dummyGame);

      dispatch('joinGame', {
        gameId: newGame.id,
        playerName: payload.hostPlayerName,
      });
    },
    /** Joins a game by setting the gameId and creating a player with the player info */
    joinGame(
      { dispatch, commit },
      payload: { gameId: string; playerName: string }
    ) {
      commit('createGameSuccess', { newGameId: payload.gameId });

      const player: Player = { name: payload.playerName };
      dispatch('addUserPlayer', player);

      dispatch('routeToLobby', { newGameId: payload.gameId });
    },
    /** Changes the gameStarted boolean so the game is started  */
    startGame({ state, dispatch, rootState }) {
      const { gameId } = state;

      dispatch('addDeck');
      if (gameId) {
        // set game started to true and set the current player id to the first player
        const firstPlayerId = rootState.player.players[0].id;
        db.collection('games')
          .doc(gameId)
          .update({ gameStarted: true, currentPlayer: firstPlayerId });
      }
    },
    /** Setup the firebase binding for the game */
    setupGameBinding: firestoreAction(async ({ bindFirestoreRef, state }) => {
      const { gameId } = state;
      if (gameId) {
        return bindFirestoreRef('game', db.collection('games').doc(gameId), {
          reset: false,
        });
      }
    }),
    routeToLobby({}, payload: { newGameId: string }) {
      router.push(`lobby/${payload.newGameId}`);
    },
    routeToGame({ state }) {
      const { gameId } = state;
      if (gameId) {
        router.push({ name: 'game', params: { gameId } });
      }
    },
  },
};
