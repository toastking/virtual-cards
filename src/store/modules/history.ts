import { db } from '@/db';
import { State } from '@/store';
import { firestore } from 'firebase';
import { Module } from 'vuex';
import { firestoreAction } from 'vuexfire';
import { Player } from './player';

/** A record of what card was drawn by what user */
export interface HistoryEntry {
  playerId: string;
  /** String representation of a card based on vue-plaing-card */
  card: string;
  timestamp: firestore.Timestamp;
}
/** State for which card was drawn and when */
export interface HistoryState {
  history: HistoryEntry[] | null;
}

export const HistoryModule: Module<HistoryState, State> = {
  state: () => ({ history: [] }),
  actions: {
    /** Adds the history to a game */
    addHistory({ rootState }, payload: { player: Player; card: string }) {
      const historyEntry: HistoryEntry = {
        playerId: payload.player?.id ?? '',
        card: payload.card,
        timestamp: firestore.Timestamp.fromDate(new Date()),
      };

      const { gameId } = rootState.game;
      if (gameId) {
        return db
          .collection('games')
          .doc(gameId)
          .collection('history')
          .add(historyEntry);
      }
    },
    setupHistoryBinding: firestoreAction(
      async ({ rootState, bindFirestoreRef }) => {
        const { gameId } = rootState.game;
        if (gameId) {
          return bindFirestoreRef(
            'history',
            db
              .collection('games')
              .doc(gameId)
              .collection('history')
              .orderBy('timestamp', 'desc')
          );
        }
      }
    ),
  },
  getters: {
    historyForPlayer(state) {
      return (player: Player) => {
        if (state.history) {
          return state.history.filter(entry => entry.playerId === player.id);
        }
      };
    },
  },
};
