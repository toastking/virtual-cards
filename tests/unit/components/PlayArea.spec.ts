import PlayArea from '@/components/PlayArea.vue';
import { DeckState } from '@/store/modules/deck';
import { createLocalVue, mount, shallowMount } from '@vue/test-utils';
import VuePlayingCard from 'vue-playing-card';
import { Store } from 'vuex-mock-store';

describe('PlayArea', () => {
  const decks: DeckState['decks'] = [{ drawnCards: [] }];

  const store = new Store({
    getters: { gameOver: false, decks, isYourTurn: true, turnIsLoading: false },
  });
  // add other mocks here so they are accessible in every component
  const mocks = {
    $store: store,
  };

  const localVue = createLocalVue();
  localVue.use(VuePlayingCard);

  afterEach(() => {
    store.reset();
  });

  describe('snapshots', () => {
    test('show only the blank deck when there are no decks', () => {
      store.getters.decks = [];
      const wrapper = shallowMount(PlayArea, { mocks, localVue });
      expect(wrapper.element).toMatchSnapshot();
    });

    test('show a card when there is one', () => {
      store.getters.decks[0].drawnCards = ['4s'];
      store.getters.decks[0].currentCard = '4s';

      const wrapper = shallowMount(PlayArea, { mocks, localVue });
      expect(wrapper.element).toMatchSnapshot();
    });

    test('hide cards when the game is over', () => {
      store.getters.gameOver = true;

      const wrapper = shallowMount(PlayArea, { mocks, localVue });
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  test('clicking the card draws a card', () => {
    const wrapper = shallowMount(PlayArea, { mocks, localVue });
    wrapper.find('.remaining-deck').trigger('click');
    expect(store.dispatch).toBeCalledWith('doTurn');
  });
});
