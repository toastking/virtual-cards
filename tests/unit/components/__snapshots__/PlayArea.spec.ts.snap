// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayArea snapshots hide cards when the game is over 1`] = `
<section>
  <div
    class="playing-card-container container box"
  >
    <div
      class="level"
    >
      <!---->
       
      <div
        class="level-item"
      >
        <vue-playing-card-stub
          enter-active-class="animate__animated animate__flipInY"
          name="playingcard"
          signature="4s"
          tag="div"
        />
      </div>
    </div>
  </div>
</section>
`;

exports[`PlayArea snapshots show a card when there is one 1`] = `
<section>
  <div
    class="playing-card-container container box"
  >
    <div
      class="level"
    >
      <div
        class="level-item remaining-deck"
      >
        <vue-playing-card-stub
          cover="true"
        />
      </div>
       
      <div
        class="level-item"
      >
        <vue-playing-card-stub
          enter-active-class="animate__animated animate__flipInY"
          name="playingcard"
          signature="4s"
          tag="div"
        />
      </div>
    </div>
  </div>
</section>
`;

exports[`PlayArea snapshots show only the blank deck when there are no decks 1`] = `
<section>
  <div
    class="playing-card-container container box"
  >
    <div
      class="level"
    >
      <div
        class="level-item remaining-deck"
      >
        <vue-playing-card-stub
          cover="true"
        />
      </div>
       
    </div>
  </div>
</section>
`;
